_posts/2019-09-20-knative.md

```diff
@@ -0,0 +1,159 @@
+
+Before developing an application there is one more step that must be completed - adding Kafka support to Knative.
+Knative Eventing comes with EventSources that can emit messages from Kubernetes platform events, GitHub events and based on time.
+The Kafka EventSource is not part of the core installation so must be added separately.
+
+```bash
+kubectl apply -f https://github.com/knative/eventing-contrib/releases/download/v0.8.0/kafka.yaml
+```
+
+# Deploying an application
+
+Knative provides a number of different ways to consume events from Apache Kafka.
+The simplest, which will be shown here, is to connect a Kafka EventSource directly to a Knative Serving application.
+Others include decoupling the Source and Service using a Channel (more applicable when the source is not a messaging platform itself) and routing via an internal Broker.
+
+![Knative Eventing Architecture]({{ "/assets/2019-09-20-knative-eventing-architecture.png" }})
+
+We will use a pre-built Knative Serving application that will log events to the console.
+The yaml for the service is shown below with the key points being the `name` and `namespace`.
+
+```yaml
+apiVersion: serving.knative.dev/v1alpha1 
+kind: Service
+metadata:
+  name: logevents
+  namespace: default 
+spec:
+  template:
+    spec:
+      containers:
+      - image: docker.io/matzew/event-display 
+```
+
+Save this yaml to a file and deploy it with the usual `kubectl apply -f service.yaml`. 
+This will create a Custom Resource for the serving function and an initial pod to service the requests.
+You can view the state of the serving function using the `ksvc` CRD.
+
+```bash
+$ kubectl get ksvc
+NAME        URL                                    LATESTCREATED     LATESTREADY       READY     REASON
+logevents   http://logevents.default.example.com   logevents-c5jlv   logevents-c5jlv   True
+```
+
+The Serving application automatically exposes an HTTP endpoint to invoke the service from inside and outside the Kubernetes cluster.
+However, we want the service to respond to messages on a Kafka Topic.
+Therefore we create a `KafkaSource` Custom Resource which controls the Topic -> application mapping.
+The `KafkaSource` includes properties for the address of the Kafka Broker, the topic to consume from and the consumer group that the consumer should belong to.
+It is also possible to configure the source to use TLS encryption and SASL authentication if required by the Kafka broker - the Source is able to extract the relevant usernames, passwords and certificates from Kubernetes Secrets.
+Finally, in the `spec.sink` section the Source references the Serving application that was created earlier.
+
+
+```yaml
+apiVersion: sources.eventing.knative.dev/v1alpha1
+kind: KafkaSource
+metadata:
+  name: kafka-source
+  namespace: default
+spec:
+  consumerGroup: knative-group
+  bootstrapServers: my-cluster-kafka-bootstrap.kafka.svc:9092
+  topics: my-topic
```

Comment:
w/o a specific yaml for the `KafkaTopic` it's one w/ just one partition.

Perhaps, weoth to mention some of the "scale to n pods" ? 

As a note or so at the bottom of the post?   I can think about something.

because the SRC has one (concurrent) `goroutine` PER partition.

As, the `autoscaling` config of ksvc: 100 concurrent reqyuest.  If you minize that: (w/ 100 partitions). you get n pods, for instance

Reply:
@matzew Better now?